galaxyproject/galaxy
# Fix revision scripts, run migrations in CI, add repair option, improve migrations utils

Builds on top of #15876.
Ref #15746, #15772

Some revision scripts ~~are currently~~ were broken, both for postgresql and sqlite. Running `./manage_db.sh downgrade base` and/or `./manage_db.sh upgrade` (from base) ~~raises~~ would raise errors. There ~~are~~ were multiple causes, all of which are resolved in this PR (see individual commits for more details)

### Notes
This builds on top of #15876. First new commit is "[Add db migrate test to postgres workflow](https://github.com/galaxyproject/galaxy/pull/15811/commits/65b2c86d39c71e41a99fbbd314b6cc4a17ae180d)".

I've added a unit test that applies all revisions (head >> base >> head >> base). [See docstring for details](https://github.com/galaxyproject/galaxy/pull/15811/files#diff-a7f0b7bb8b12327e91eaaa702276d920064f81cad19248afae65e9a399b2fabfR27)
The test will be executed as part of the `unit-postgres.yaml` github workflow, twice: on sqlite and postgres (locally it'll run on sqlite and on postgres if it's available). In addition to verifying that all revisions can be applied w/o error, this test will also fail when a revision has been added, but the model has not been updated (provided the downgrade script reverses the changes specified in the upgrade script).

I've added a `--repair` option for the `db_dev` script's upgrade/downgrade commands. Setting the option will skip revisions that conflict with the state of the database ([see help text](https://github.com/galaxyproject/galaxy/pull/15811/files#diff-20863783418945e422d31e8db9e120a6e4c2508e9e65d13bc533144c7cda26deR163)). The option is *not* available on the user-facing `manage_db.sh` script.

The [`transaction` context manager](https://github.com/galaxyproject/galaxy/pull/15811/files#diff-ee19f5fa9ace799394cb0f7f35ea3a70e5dc62e7c96899e795e39cebf81e17e1R353) wraps multiple DDL statements ensuring transactional control when running under SQLite (more context: https://bugs.python.org/issue10740)

The [`legacy_alter_table` context manager](https://github.com/galaxyproject/galaxy/pull/15811/files#diff-ee19f5fa9ace799394cb0f7f35ea3a70e5dc62e7c96899e795e39cebf81e17e1R338) prevents the error raised when a column belongs to a table referenced in a database view (without this, we can't downgrade to base on SQLite). (more context: https://github.com/sqlalchemy/alembic/issues/1207)

## How to test the changes?
- [x] I've included *some* [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html). More tests are needed for the `model.migrations.util` module; however, those require additional testing infrastructure that's outside of the scope of this PR.
 1. Start with an up-to-date database. 
 2. Run `./manage_db.sh downgrade base; ./manage_db.sh upgrade`. Observe no error.
 3. Comment out the statements in a revision's upgrade OR downgrade function (I tried revision 6a67bf27e6a6)
 4. Repeat step 2,.*Depending on what you commended out, you may need to run the commands TWICE to trigger the error*. Observe the error.
 5. Uncomment what you commented in state 3. Repeat step 4 - you still have an error - the database is broken! 
*No need to fear, `--repair` is here!*
 6. Run `/scripts/db_dev.sh downgrade base --repair; /scripts/db_dev.sh upgrade --repair`.
 7. Now repeat step 2. Observe no error. You've fixed your database!
 8. Repeat all steps using postgresql or sqlite, so you try both databases.

Note: For step 2, I didn't try breaking all revisions, so it still might be possible to break things beyond what `--repair` can handle. But it does handle the obvious/typical cases.

## License
